File: sir_comp.py
#!/usr/bin/env python

import scipy.integrate as spi
import numpy as np
import pylab as pl

beta = 0.5 ## infection rate
gamma = 0.05 ## recovery rate

t_inc = 1 ## increament in the number of days 
num_days = 200 ## total number of days to track the infection

## Here it is calculated for the fraction of people. It can be changed to the number of people.
## To change, make N_total = total number of people and I0 = the initial number of infected people
## For example, N_total = 1e6, I_0 = 1.
## This means the total size of the population is 1 million where initially only 1 person is infected.

N_total = 1           ## total population size

I0 = 1e-5             ## initial fraction of infected people
S0 = N_total - I0     ## initial fraction of susceptible people
R0 = 0                ## initial fraction of recovered people

start = (S0, I0, R0)


def diff_eqs(init, t):  
	'''The main set of equations'''
    
	Y = np.zeros((3))     ## initialize the Y-vector with zeros
	V = init              
    
	Y[0] = - beta * V[0] * V[1]
	Y[1] = beta * V[0] * V[1] - gamma * V[1]
	Y[2] = gamma * V[1]
	return Y   # For odeint

t_start = 0.0; t_end = num_days; t_inc = t_inc

t_range = np.arange(t_start, t_end+t_inc, t_inc)

result = spi.odeint(diff_eqs, start, t_range)

print(result) 

## in python 2, print result
## in python 3, print(result)


#Ploting
pl.figure(figsize=(10,10))   ## define figure size

axis_font = {'size':'35', 'color':'black'}
axis_tick_font = {'size':'25', 'color':'black'}

pl.plot(result[:,0], '-b', label='Susceptibles', lw = 3)
pl.plot(result[:,1], '-r', label='Infectious', lw = 3)
pl.plot(result[:,2], '-g', label='Recovered', lw = 3)
pl.legend(loc=0, fontsize = 25)
pl.xlabel('Time', **axis_font)
pl.ylabel('Fraction of People', **axis_font)

pl.xticks(**axis_tick_font)
pl.yticks(**axis_tick_font)

pl.savefig('C:/Users/nmaddy/Documents/git/epi_model/sir.png')

pl.show()

File: sircomp.R
##install packages
install.packages('deSolve')

## Load deSolve package
library(deSolve)

## Create an SIR function
sir <- function(time, state, parameters) {

  with(as.list(c(state, parameters)), {

    dS = -beta * S * I
    dI = beta * S * I - gamma * I
    dR = gamma * I

    return(list(c(dS, dI, dR)))
  })
}

### Set parameters
## Fraction in each compartment: 
## Total population size = N = 1, Infected = 1*10^-5, Susceptibles = N - I, Recovered = 0
N_total = 1
I0 = 1e-05
S0 = N_total - I
R0 = 0.0
init = c(S = S0, I = I0, R = R0)

## beta: infection rate; gamma: recovery rate
parameters = c(beta = 0.5, gamma = 0.05)
## Time frame
days = seq(0, 200, by = 1)

## Solve using ode (General Solver for Ordinary Differential Equations)
out = ode(y = init, times = days, func = sir, parms = parameters)

## change to data frame
out = as.data.frame(out)

## Show data
head(out, 10)

out$time = NULL
## remove the time column from the dataframe, to make it easier to plot


## Plot
fig = matplot(x = days, y = out, type = "l",
        xlab = "Time", ylab = "Fraction of People",
        lwd = 3, lty = 1, bty = "l", col = 2:4)
box(which = "plot")

## Add legend
legend(140, 0.7, c("Susceptible", "Infected", "Recovered"),
       pch = c(NA, NA, NA), lty = 1, col = 2:4, bty = "o", lwd = 3)



